Sources/SotoCore/HTTP/S3StreamWriter.swift

```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Soto for AWS open source project
+//
+// Copyright (c) 2017-2020 the Soto project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Soto project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import SotoSignerV4
+
+class S3StreamWriter: StreamWriterProtocol {
+    /// Working buffer size
+    static let bufferSize: Int = 64 * 1024
+    static let bufferSizeInHex = String(bufferSize, radix: 16)
+    static let bufferSizeLength = bufferSizeInHex.count
+    static let chunkSignatureLength = 1 + 16 + 64 // ";" + "chunk-signature=" + hex(sha256)
+    static let endOfLineLength = 2
+    /// Maximum size of chunk header
+    static let maxHeaderSize: Int = bufferSizeInHex.count + chunkSignatureLength + endOfLineLength
+
+    let length: Int
+    var bytesWritten: Int
+    let eventLoop: EventLoop
+    let writerPromise: EventLoopPromise<ChildStreamWriter>
+    let finishedPromise: EventLoopPromise<Void>
+
+    /// bytebuffer allocator
+    var byteBufferAllocator: ByteBufferAllocator
+
+    var signer: AWSSigner
+    var signingData: AWSSigner.ChunkedSigningData
+    var headerBuffer: ByteBuffer
+    var workingBuffer: ByteBuffer
+    var tailBuffer: ByteBuffer
+
+    init(
+        length: Int,
+        signer: AWSSigner,
+        seedSigningData: AWSSigner.ChunkedSigningData,
+        byteBufferAllocator: ByteBufferAllocator,
+        eventLoop: EventLoop
+    ) {
+        self.length = Self.calculateLength(from: length)
+        self.bytesWritten = 0
+        self.eventLoop = eventLoop
+        self.writerPromise = eventLoop.makePromise()
+        self.finishedPromise = eventLoop.makePromise()
+
+        self.signer = signer
+        self.signingData = seedSigningData
+        // have separate buffers so we aren't allocating 128k
+        self.byteBufferAllocator = byteBufferAllocator
+        self.headerBuffer = byteBufferAllocator.buffer(capacity: Self.maxHeaderSize)
+        self.workingBuffer = byteBufferAllocator.buffer(capacity: Self.bufferSize)
+        self.tailBuffer = byteBufferAllocator.buffer(capacity: Self.endOfLineLength)
+        self.tailBuffer.writeString("\r\n")
+    }
+
+    /// Update HTTP headers. Add "Content-encoding" header. The "x-amz-decoded-content-length" header is added earlier in AWSRequest when
+    /// initiating the signing process
+    /// - Parameter headers: headers to update
+    func updateHeaders(headers: HTTPHeaders) -> HTTPHeaders {
+        var headers = headers
+        headers.add(name: "Content-Encoding", value: "aws-chunked")
```

Comment:
nit - headers _should_ be lowercased (at least in HTTP/1 - it's a must in HTTP/2 IIRC) and are lowercased internally so I suggest we follow that. Thoughts?

```suggestion
        headers.add(name: "content-Encoding", value: "aws-chunked")
```

Reply:
If we're gonna lowercase the "c" we should probably also do the same to the "e" 😝